components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ServerURL.java

```diff
@@ -0,0 +1,411 @@
+                    serverUrl.append("/t/").append(tenantDomain);
+                }
+            }
+        }
+
+        if (StringUtils.isBlank(resolvedUrlContext)) {
+            resolvedUrlContext = resolveUrlFromServerConfig(urlContext);
+        }
+
+        if (StringUtils.isNotBlank(resolvedUrlContext)) {
+            appendURLContext(serverUrl, resolvedUrlContext);
+        }
+
+        if (!isTenantQualifiedUrlsEnabled() && this.getAppendTenantAsQueryParam()) {
+            if (StringUtils.isNotBlank(tenantDomain) &&
+                    !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equalsIgnoreCase(tenantDomain)) {
+                serverUrl.append("?").append(MultitenantConstants.TENANT_DOMAIN).append("=").append(tenantDomain);
+            }
+        }
+
+        if (serverUrl.toString().endsWith("/")) {
+            serverUrl.setLength(serverUrl.length() - 1);
+        }
+    }
+
+    /**
+     * Server URL builder representation.
+     */
+    public static class ServerURLBuilder {
+
+        private String protocol;
+        private String hostName;
+        private Integer port;
+        private String proxyContext;
+        private String webContext;
+        private Boolean addProxyContext;
+        private Boolean addWebContext;
+        private String urlContext;
+        private Boolean appendTenantAsQueryParam;
+        private Boolean appendTenantAsPathParam;
+        private ArrayList<String> parameters;
+        private ArrayList<String> fragments;
+
+        public ServerURLBuilder(String urlContext) {
+
+            this.urlContext = urlContext;
+        }
+
+        public ServerURLBuilder protocol(String protocol) {
+
+            this.protocol = protocol;
+            return this;
+        }
+
+        public ServerURLBuilder hostName(String hostName) {
+
+            this.hostName = hostName;
+            return this;
+        }
+
+        public ServerURLBuilder port(Integer port) {
+
+            this.port = port;
+            return this;
+        }
+
+        public ServerURLBuilder proxyContext(String proxyContext) {
+
+            this.proxyContext = proxyContext;
+            return this;
+        }
+
+        public ServerURLBuilder webContext(String webContext) {
+
+            this.webContext = webContext;
+            return this;
+        }
+
+        public ServerURLBuilder addProxyContext() {
+
+            this.addProxyContext = true;
+            return this;
+        }
+
+        public ServerURLBuilder addWebContext() {
+
+            this.addWebContext = true;
+            return this;
+        }
+
+
+        public ServerURLBuilder appendTenantAsQueryParam() {
+
+            this.appendTenantAsQueryParam = true;
+            return this;
+        }
+
+        public ServerURLBuilder appendTenantAsPathParam() {
+
+            this.appendTenantAsPathParam = true;
+            return this;
+        }
+
+        public ServerURLBuilder parameters(ArrayList<String> parameters) {
```

Comment:
Also would be better to have a parameter() method that allows adding params one by one.
Also shouldn't the input be a map?